syncer/parallel_sync.go

```diff
@@ -0,0 +1,201 @@
+		base := types.ChainIndex{
+			ID:     headers[off].ParentID,
+			Height: cs.Index.Height + off,
+		}
+		tip := types.ChainIndex{
+			ID:     headers[off+numBlocks-1].ID(),
+			Height: base.Height + numBlocks - 1,
+		}
+		reqs[i] = Req{base, tip, numBlocks}
+	}
+
+	workFn := func(p *Peer, req Req) (resp Resp) {
+		resp.req = req
+		startTime := time.Now()
+		if req.base.Height >= cs.Network.HardforkV2.RequireHeight {
+			cs, b, err := p.SendCheckpoint(req.base, cs.Network, s.config.SendBlockTimeout)
+			if err != nil {
+				return Resp{req: req, err: err}
+			}
+			cs, _ = consensus.ApplyBlock(cs, b, consensus.V1BlockSupplement{}, time.Time{})
+			blocks, _, err := p.SendV2Blocks(ctx, []types.BlockID{cs.Index.ID}, req.numBlocks, s.config.SendBlocksTimeout)
+			if err != nil {
+				return Resp{req: req, err: err}
+			} else if uint64(len(blocks)) != req.numBlocks {
+				return Resp{req: req, err: errors.New("peer returned wrong number of blocks")}
+			} else if blocks[len(blocks)-1].ID() != req.tip.ID {
+				return Resp{req: req, err: errors.New("peer returned wrong blocks")}
+			}
+			resp.blocks = blocks
+			for _, b := range blocks {
+				if err := consensus.ValidateBlock(cs, b, consensus.V1BlockSupplement{}); err != nil {
+					return Resp{req: req, err: err}
+				}
+				cs, _ = consensus.ApplyBlock(cs, b, consensus.V1BlockSupplement{}, time.Time{})
+				resp.states = append(resp.states, cs)
+			}
+		} else {
+			blocks, _, err := p.SendV2Blocks(ctx, []types.BlockID{req.base.ID}, req.numBlocks, s.config.SendBlocksTimeout)
+			if err != nil {
+				return Resp{req: req, err: err}
+			} else if uint64(len(blocks)) != req.numBlocks {
+				return Resp{req: req, err: errors.New("peer returned wrong number of blocks")}
+			} else if blocks[len(blocks)-1].ID() != req.tip.ID {
+				return Resp{req: req, err: errors.New("peer returned wrong blocks")}
+			}
+			resp.blocks = blocks
+		}
+		endTime := time.Now()
+		s.pm.UpdatePeerInfo(p.t.Addr, func(info *PeerInfo) {
+			info.SyncedBlocks += req.numBlocks
+			info.SyncDuration += endTime.Sub(startTime)
+		})
+		return
+	}
+
+	reqChan := make(chan Req, 128) // surely a reasonable maximum number of peers
+	respChan := make(chan Resp, 128)
+
+	// process results in a separate goroutine
+	finishCh := make(chan Resp, len(reqs))
+	errCh := make(chan error, 1)
+	resps := make([]*Resp, len(reqs))
+	go func() {
+		i := 0
+		for i < len(resps) {
+			r, ok := <-finishCh
+			if !ok {
+				break
+			}
+			off := (r.req.base.Height - cs.Index.Height) / blocksPerReq
+			if resps[off] != nil {
+				continue // already completed by a faster worker
+			}
+			resps[off] = &r
+			for i < len(resps) && resps[i] != nil {
+				r := resps[i]
+				var err error
+				if r.req.base.Height >= cs.Network.HardforkV2.RequireHeight {
+					err = s.cm.AddValidatedV2Blocks(r.blocks, r.states)
+				} else {
+					err = s.cm.AddBlocks(r.blocks)
+				}
+				if err != nil {
+					errCh <- err
```

Comment:
Can this ever happen outside of a persistence error? My concern here is that an invalid block breaks syncing.

Reply:
immediately after the `SendV2Blocks` calls